### PR TITLE
testing strong commit

### PR DESCRIPTION
**Pull request by danielxiangzl**
implement strong vote with marker

add marker info to the QC

implemented the computation of the number of endorsers

## Motivation

(Write your motivation for proposed changes here.)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

## Test Plan

(Share your test plan here. If you changed code, please provide us with clear instructions for verifying that your changes work.)

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/website, and link to your PR here.)

### CODE REVIEW

File: consensus/consensus-types/src/block_data.rs [hunk 1]
```diff
@@ -148,6 +148,7 @@ where
             LedgerInfoWithSignatures::new(
                 LedgerInfo::new(ancestor, HashValue::zero()),
                 BTreeMap::new(),
+                BTreeMap::new(),
```
Comment (dahliamalkhi) [inline consensus/consensus-types/src/block_data.rs:151]: I am not sure we need to aggregate markers in the QC. 

It seems we may want to aggregate endorsements on a target block, coming from votes on potentially different blocks, whose markers endorse the target.

File: consensus/consensus-types/src/timeout_certificate.rs [hunk 2]
```diff
@@ -17,6 +17,7 @@ use std::{collections::BTreeMap, fmt};
 pub struct TimeoutCertificate {
     timeout: Timeout,
     signatures: BTreeMap<Author, Ed25519Signature>,
+    markers: BTreeMap<Author, Round>,
```
Comment (dahliamalkhi) [inline consensus/consensus-types/src/timeout_certificate.rs:20]: really puzzled: why do we need markers on timeout certificates?

File: consensus/consensus-types/src/vote.rs [hunk 3]
```diff
@@ -27,17 +28,20 @@ pub struct Vote {
     signature: Ed25519Signature,
     /// The round signatures can be aggregated into a timeout certificate if present.
     timeout_signature: Option<Ed25519Signature>,
+    /// The marker equals the largest round number of any conflicting block that the validator voted for
+    marker: Round,
```
[the remark this answers is not in the document]
Comment (danielxiangzl) [reply consensus/consensus-types/src/vote.rs:31]: Agreed.

File: consensus/consensus-types/src/block_data.rs (hunk 1 above)
Comment (danielxiangzl) [reply consensus/consensus-types/src/block_data.rs:151]: The current LibraBFT implementation does not use threshold sig. Instead, QC stores 2f+1 signatures in the BTreeMap struct. I am doing similar things for the marker, but the current implementation is not secure (markers are not signed).

File: consensus/consensus-types/src/timeout_certificate.rs (hunk 2 above)
Comment (danielxiangzl) [reply consensus/consensus-types/src/timeout_certificate.rs:20]: Yes, you are right..we don't need markers on TC. Earlier I thought a TC can serve as a QC as well, but it is not the case in the current implementation.

**Comment by davidiw** [comment]
Curious if this is intended to be a feature added to the Libra code base or an exploratory project? If the latter, I would recommend doing this as a PR in your own repo. We've done the similar with our exploration of SGX: https://github.com/zaxguo/libra/pull/1

**Comment by danielxzl** [comment]
> Curious if this is intended to be a feature added to the Libra code base or an exploratory project? If the latter, I would recommend doing this as a PR in your own repo. We've done the similar with our exploration of SGX: [zaxguo#1](https://github.com/zaxguo/libra/pull/1)

Hi, this is currently a research project and hopefully can be added to the codebase in the future :) The PR is mainly for the purpose of cluster test and measurements, I would be happy to close this and reopen it only when doing the measurement.

**Pull request closed without merging**